File: lib/json_api/utterance.rb
module JsonApi::Utterance
  extend JsonApi::Json
  
  TYPE_KEY = 'utterance'
  DEFAULT_PAGE = 10
  MAX_PAGE = 25
    
  def self.build_json(utterance, args={})
    json = {}
    json['id'] = utterance.global_id
    json['created_at'] = utterance.created_at.iso8601
    json['button_list'] = utterance.data['button_list']
    json['link'] = "#{JsonApi::Json.current_host}/utterances/#{utterance.global_id}"
    json['sentence'] = utterance.data['sentence']
    json['image_url'] = utterance.data['image_url'] || "https://s3.amazonaws.com/opensymbols/libraries/noun-project/Person-08e6d794b0.svg"
    json['large_image_url'] = utterance.data['large_image_url']
    if args[:permissions]
      json['permissions'] = utterance.permissions_for(args[:permissions])
    end
    if ((json['permissions'] && json['permissions']['edit']) || utterance.data['show_user']) && utterance.user
      json['user'] = JsonApi::User.as_json(utterance.user, limited_identity: true)
    end
    json['show_user'] = !!utterance.data['show_user']
    json
  end
end

File: app/frontend/app/models/utterance.js
import Ember from 'ember';
import DS from 'ember-data';
import CoughDrop from '../app';

CoughDrop.Utterance = DS.Model.extend({
  button_list: DS.attr('raw'),
  sentence: DS.attr('string'),
  link: DS.attr('string'),
  image_url: DS.attr('string'),
  large_image_url: DS.attr('string'),
  permissions: DS.attr('raw'),
  user: DS.attr('raw'),
  show_user: DS.attr('boolean'),
  best_image_url: function() {
    return this.get('large_image_url') || this.get('image_url');
  }.property('image_url', 'large_image_url'),
  check_for_large_image_url: function() {
    var attempt = this.get('large_image_attempt') || 1;
    var _this = this;
    if(_this.get('permissions.edit') && !_this.get('large_image_url') && attempt < 15) {
      Ember.run.later(function() {
        _this.set('large_image_attempt', attempt + 1);
        _this.reload().then(function(u) {
          _this.check_for_large_image_url();
        });
      }, attempt * 500);
    }
  },
});

export default CoughDrop.Utterance;
